framework/py/flwr/supernode/nodestate/nodestate.py

```diff
@@ -15,17 +15,143 @@
 """Abstract base class NodeState."""
 
 
-import abc
-from typing import Optional
+from abc import ABC, abstractmethod
+from collections.abc import Sequence
+from typing import Optional, Union
 
+from flwr.common import Context, Message
+from flwr.common.typing import Run
 
-class NodeState(abc.ABC):
-    """Abstract NodeState."""
 
-    @abc.abstractmethod
-    def set_node_id(self, node_id: Optional[int]) -> None:
+class NodeState(ABC):
+    """Abstract base class for node state."""
+
+    @abstractmethod
+    def set_node_id(self, node_id: int) -> None:
         """Set the node ID."""
 
-    @abc.abstractmethod
+    @abstractmethod
     def get_node_id(self) -> int:
         """Get the node ID."""
+
+    @abstractmethod
+    def store_message(self, message: Message) -> Optional[str]:
+        """Store a message.
+
+        Parameters
+        ----------
+        message : Message
+            The message to store.
+
+        Returns
+        -------
+        Optional[str]
+            The object ID of the stored message, or None if storage failed.
+        """
+
+    @abstractmethod
+    def get_message(
+        self,
+        *,
+        run_ids: Optional[Union[int, Sequence[int]]] = None,
+        is_reply: Optional[bool] = None,
+        limit: Optional[int] = None,
+    ) -> Sequence[Message]:
+        """Retrieve messages based on the specified filters.
+
+        If a filter is set to None, it is ignored.
+        If multiple filters are provided, they are combined using AND logic.
+
+        Parameters
+        ----------
+        run_ids : Optional[Union[int, Sequence[int]]] (default: None)
+            Run ID or sequence of run IDs to filter by. If a sequence is provided,
+            it is treated as an OR condition.
+        is_reply : Optional[bool] (default: None)
+            If True, filter for reply messages; if False, filter for non-reply
+            (instruction) messages.
+        limit : Optional[int] (default: None)
+            Maximum number of messages to return. If None, no limit is applied.
+
+        Returns
+        -------
+        Sequence[Message]
+            A sequence of messages matching the specified filters.
+
+        Notes
+        -----
+        **IMPORTANT:** Retrieved messages will **NOT** be returned again by subsequent
+        calls to this method, even if the filters match them.
+        """
+
+    @abstractmethod
+    def delete_message(
+        self,
+        *,
+        message_ids: Optional[Union[str, Sequence[str]]] = None,
```

Comment:
Should we simplify this to:
```suggestion
        message_ids: Optional[Sequence[str]] = None,
```